### PR TITLE
Chemoreceptor can detect cells

### LINKED ISSUE #3072: Chemoreceptor should have option to detect other cells
some kind of selection for specific species or like predator or prey (based on some criteria).

### PR DESCRIPTION
**Pull request by ottoblep**
**Brief Description of What This PR Does**

This allows chemoreceptors to detect other cells.

**Related Issues**

This closes #3072.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by ottoblep** [comment]
Tracking of cells is based on species similar to compound type. The closest member of the selected species is shown.
Functionally this works almost the exact way the compound detection does.
The only structural difference is MicrobeSystem.FindSpeciesNearPoint performing the search instead of CompoundCloudSystem.FindCompoundNearPoint.
https://github.com/Revolutionary-Games/Thrive/blob/8189a71b65129b8faa23a881170308f08983f0dd/src/microbe_stage/MicrobeSystem.cs#L67
Since tracked microbes move, they need a faster update rate than compound lines.
That's why Compound and Species detection are currently completely separated and can be tweaked independently.
https://github.com/Revolutionary-Games/Thrive/blob/8189a71b65129b8faa23a881170308f08983f0dd/simulation_parameters/Constants.cs#L668-L669

TODO:
I did not find a way to access the current species list from the chemoreceptor update gui. 
The species selection from the gui is incomplete at the moment. 
I know little about the gui or localization system so far so i mostly left this out.
https://github.com/Revolutionary-Games/Thrive/blob/8189a71b65129b8faa23a881170308f08983f0dd/src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs#L54-L55
The player should be able to select between a compound or species to be tracked.
For testing purposes all species are currently added to detection for every cell.
https://github.com/Revolutionary-Games/Thrive/blob/8189a71b65129b8faa23a881170308f08983f0dd/src/microbe_stage/Microbe.cs#L984-L990

**Comment by hhyyrylainen** [comment]
> Since tracked microbes move, they need a faster update rate than compound lines.

This should use the same approach as for other cell tracking, the tracking is updated each frame but new items are only detected a few times per second. This way the lines are always up to date but the expensive search doesn't run each frame.

>  I did not find a way to access the current species list from the chemoreceptor update gui.

The upgrade GUI should probably be modified to be given the CurrentGame when starting so that it can look at the relevant info in there. Why specifically the CurrentGame is that I think there's possibility of needing more info than just a species list so this general rework will be better for the future.

**Comment by ottoblep** [comment]
A save upgrader / save incompatability point might be needed since chemoreceptor now has a different configuration.

**Comment by hhyyrylainen** [comment]
If the data is carefully laid out so that the old data exists, but a new nullable field for species to detect is added, the data change should not require any save data upgrading.

**Comment by ottoblep** [comment]
I tested the save compatability. Everything seems to transfer fine except for the line updates being broken until the next editor phase. Thats due to this one variable being renamed. I don't know if it's worth introducing a save upgrader for this, but can do if necessary.
https://github.com/Revolutionary-Games/Thrive/blob/05c128a8e664ae72a16f3523780bea640ea9300c/src/microbe_stage/MicrobeStage.cs#L622

**Comment by hhyyrylainen** [comment]
Why was the rename necessary? If the rename is necessary, it is actually possible to specify an overridden name to be used in JSON, that way an updated variable name can still load the old data.

**Comment by ottoblep** [comment]
The rename is necessary. It was called ```OnCompoundChemoreceptionInfo``` before and now it handles both kinds. 
I just saw the upgrader would need to update the value as well though, since the method that is referenced has changed.

**Comment by hhyyrylainen** [comment]
Okay that makes sense. 
You can use that JSON attribute to set the JSON field name to be the same as it was before to avoid the save breakage.

**Comment by hhyyrylainen** [comment]
I will perform a repository history rewrite on the Thrive repo on the 26th of June. This will cause a small disruption for PRs, which is possible to avoid if the PR can be merged before that.

For more info see: https://forum.revolutionarygamesstudio.com/t/rewriting-the-repository-history-to-remove-large-binary-files/991

**Comment by hhyyrylainen** [comment]
Any progress on this? The next release is definitely going to be a planned save breakage point, so any PRs that make it into the next release won't need to take save compatibility into account at all.

**Comment by ottoblep** [comment]
> Any progress on this? The next release is definitely going to be a planned save breakage point, so any PRs that make it into the next release won't need to take save compatibility into account at all.

The save upgrader is the only thing holding this up. So if the next release is a breakage point this is ready to merge. I will do some more testing and then check the boxes.

**Comment by hhyyrylainen** [comment]
There was that master branch rewrite so this needs to be remade on a clean new branch, all the commits cherry-picked from the old branch and then pushed over this to make this mergeable.

**Comment by ottoblep** [comment]
> There was that master branch rewrite so this needs to be remade on a clean new branch, all the commits cherry-picked from the old branch and then pushed over this to make this mergeable.

I can do that. This needs a new pull request then right?

**Comment by hhyyrylainen** [comment]
You can just force push to the same branch to clear out the old stuff.

**Comment by hhyyrylainen** [comment]
Actually I had a locally already fixed up branch of that. That just leaves the translation files conflicting. So if you first force pull you should have a *much* easier time to fix this branch with that as the base.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:8]: Only the first NodePath should be nullable. That's the style used in Thrive. So the other NodePaths here as well as the Dispose method need to be slightly modified to match Thrive style.

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:908]: This is an incorrect change.

File: src/microbe_stage/MicrobeSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSystem.cs:75]: This method should not rebuild the microbe species list each time this is called... instead the caller should specify it or maybe even the microbe system could cache an internal list of microbes in the Process method in this file?

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:912]: Is Visible a nullable type? If not it should not be explicitly compared against a boolean, that's also not the Thrive style.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.tscn
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.tscn:27]: As the options are static, they should be defined here in the scene instead of the .cs file. There's a special button in the Godot editor to edit the contents of an OptionButton.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:82]: Why not compare against formatted name? That way the second part of the name would also help in making sure things are alphabetical.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:86]: Manually building the species name display is not correct usage. As I said in my previous comment the species class already has a property for the properly formatted name.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:90]: `var` should be preferred.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:124]: Seeing as 0 and 1 are used as magic numbers in multiple places, I think you should specify an enum like in other places where different tabs are used in the GUI to add some bulletproofing to the code.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:207]: Does this make a sensible selection when the species is extinct and is no longer found? My hunch is that it does, but I just want to ask to confirm.

File: src/microbe_stage/Microbe.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.cs:984]: Oh and this is called here in a recursive loop. This is absolutely going to wreck performance if you have a colony of 50 cells each with multiple chemoreceptor and hundreds of cells in the game, the full list of microbes in the world is going to be rebuilt so many times in that situation.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I ended up doing a code review piecemeal so see my previous comments. I didn't playtest this yet as I identified a few potential performance problems and some general code style problems.

File: src/microbe_stage/MicrobeStage.cs
Comment (ottoblep) [reply src/microbe_stage/MicrobeStage.cs:912]: Visible is not nullable here. Which style would you use? I couldn't find a definitive answer in the existing code or style guide.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs:912]: `if (!chemoreception.Line.Visible)`

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (ottoblep) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:221]: Yes, since the configuration saves the species/compound itself this should be robust against changes in the world. On entering the editor if the species is no longer found it sets the `OptionButton` to unselected. If it stays unselected it is set to the first species in the list on `ApplyChanges`. If the editor stage is skipped it is possible that an extinct species stays in the configuration. Then the component simply does nothing until edited again.

Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:221]: That sounds about what I expected would happen, which is a fine approach.

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:921]: This is not accurate. ColonyParent technically isn't required t obe the colony lead cell. As you actually want the colony lead cell you need to go through the MicrobeColony object and grab the master from there.

**Comment by ottoblep** [comment]
Is there a standard way to merge these translation files? It seems they can get easily messed up when the line numbers change.

**Comment by hhyyrylainen** [comment]
I just accept the other side for all files except en.po which I merge manually. After that I rerun the update script and *then* commit the merge.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:81]: Isn't there a "cleaner" sort overload something like `shownCompoundChoices.Sort(i => i.Name, StringComparer.OrdinalIgnoreCase)`? That avoids any potential bugs in manually performing the compare instead of just telling the standard sort algorithm to use a specific property.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:89]: And same here.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:98]: This object should only be created when necessary, i.e. in the else branch.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:129]:
```suggestion
        Compound? compoundChoice = null;
```
The word order sounds unnatural here, so here's a suggestion to improve it. Also for the next line the same thing can be done.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:8]: I'd like these added controls to have slightly more expressive names. Right now this is just "target types", which I'd expect to be a list of target types, when in fact this is the target type selection. So I suggest the name `TargetTypeSelectorPath` and renaming the related variables.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:17]:
```suggestion
    public NodePath SpeciesSelectorPath = null!;
```

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:11]:
```suggestion
    public NodePath CompoundSelectorPath = null!;
```

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:39]: And these are the related variables that should also be renamed if my previous rename comments are accepted.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:239]:
```suggestion
        if (targetTypes.Selected == (int)TargetType.Compound && shownCompoundChoices != null
            && compounds.Selected >= 0)
        {
            colour.Color = shownCompoundChoices[compounds.Selected].Colour;
        }

        if (targetTypes.Selected == (int)TargetType.Species && shownSpeciesChoices != null
            && species.Selected >= 0)
        {
            colour.Color = shownSpeciesChoices[species.Selected].Colour;
        }
```

Multiline conditions should use braces for readability purposes.

File: src/microbe_stage/MicrobeSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSystem.cs:97]: Does this *actually* perform better? This is multiple branching operations whereas the distance check should be *just* about 20 float operations and a single comparison.

File: src/microbe_stage/MicrobeSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSystem.cs:110]: These could be maybe swapped around to simplify the logic here. It probably won't perform worse as the microbes are very likely not in a descending order of distance to the target position. And it would entirely remove the need to check if the search radius is followed here. That can be added just at the end to return null instead of the closest microbe if it is farther than the search radius.

File: src/microbe_stage/MicrobeSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeSystem.cs:78]:
```suggestion
        var searchRadiusSquared = searchRadius * searchRadius;

        microbes ??= worldRoot.GetTree().GetNodesInGroup(Constants.RUNNABLE_MICROBE_GROUP).Cast<Microbe>()
            .ToArray();
```

Just a bit extra blank space to separate logically separate variables, which looks better in my opinion. You don't have to accept this suggestion.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (ottoblep) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:81]: An alternative would be `shownCompoundChoices = shownCompoundChoices.OrderBy(x=>x.Name).ToList();`, but this is slower since it allocates a new list. With a quick search i couldn't find a way to shorten the `Comparison<T>`.

Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:81]: Hang on, there's a bug here *anyway* as you are modifying the list that is owned by SimulationParameters. You should not do that. So the actual solution here is:

```
        shownCompoundChoices = SimulationParameters.Instance.GetCloudCompounds()
            .OrderBy(c => c.Name, StringComparer.OrdinalIgnoreCase).ToList();
```
(I had forgotten that the nice syntax is only supported by the LINQ sort and not the inbuilt list sort)

I'll make a separate change to make the simulation parameters return a unmodifiable list here.

File: src/microbe_stage/MicrobeSystem.cs
Comment (ottoblep) [reply src/microbe_stage/MicrobeSystem.cs:97]: Ignoring data instructions and if multiplication takes 2 ops, this roughly saves 7 ops per miss and adds 6 ops per hit. Since the vast majority of calls will be misses i think it is faster. However the difference might be negligible.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeSystem.cs:97]: Isn't any branching operation slower than even division? That's my point that optimizing a bit of math is always slower on a modern CPU as branching is so expensive operation (hence why branchless programming is such a popular performance topic).

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:89]: As [discovered](https://github.com/Revolutionary-Games/Thrive/pull/4261/files#r1313798649), there's probably no better comparison method syntax available here.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs:39]: The above variables were renamed, but these weren't yet to match the new names of the related variables.

File: src/microbe_stage/MicrobeSystem.cs
Comment (ottoblep) [reply src/microbe_stage/MicrobeSystem.cs:98]: The branch prediction will likely prefetch the `continue` path since it is the most likely. Then we need to add the cost of flushing the pipline to the added cost on a hit (lets say 10 cycles). So the change in cycles when using the branching is\
`(1-hitrate) * -7  + hitrate * 16` which is negative for a hitrate smaller than 30%. Though some profiling is probably needed to get a real comparison.

Comment (ottoblep) [reply src/microbe_stage/MicrobeSystem.cs:98]: tldr I think the branch is pretty deterministic and thus will have only a small performance impact.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeSystem.cs:98]: Well I'd be really interested to show what a benchmark done using https://github.com/dotnet/BenchmarkDotNet (I've seen this used by serious C# performance related Stackoverflow answers use) would show in this case.

File: src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs]: This change is now fine. And for reference here's the other PR I created: https://github.com/Revolutionary-Games/Thrive/pull/4454

File: src/microbe_stage/MicrobeSystem.cs
Comment (ottoblep) [reply src/microbe_stage/MicrobeSystem.cs:98]: So i built the [test bench](https://github.com/ottoblep/thrive-profiling/blob/master/Program.cs) (BenchmarkDotNet works really well) and got these results. If the search radius is large compared to the spawn radius and if there are many of the searched species then the NoBranch option is better otherwise the reverse is true. In the average in-game case i got from testing in the editor (FRAC=0.05, RADIUS=150) the Branch option is slightly better.
```
|                       Method | SPECIES_FRACTION | SEARCH_RADIUS  |     Mean |    Error |   StdDev |   Median |
|----------------------------- |----------------- |--------------- |---------:|---------:|---------:|---------:|
|   FindSpeciesNearPointBranch |             0.05 |            50  | 617.9 ns |  1.49 ns |  1.32 ns | 618.1 ns |
| FindSpeciesNearPointNoBranch |             0.05 |            50  | 663.6 ns | 19.81 ns | 54.24 ns | 645.8 ns |
|   FindSpeciesNearPointBranch |             0.05 |           150  | 638.7 ns | 12.24 ns | 14.09 ns | 634.6 ns |
| FindSpeciesNearPointNoBranch |             0.05 |           150  | 647.1 ns | 12.72 ns | 22.62 ns | 637.6 ns |
|   FindSpeciesNearPointBranch |             0.05 |           350  | 642.0 ns | 12.64 ns | 17.72 ns | 635.3 ns |
| FindSpeciesNearPointNoBranch |             0.05 |           350  | 647.6 ns | 12.91 ns | 21.57 ns | 645.9 ns |
|   FindSpeciesNearPointBranch |              0.2 |            50  | 629.8 ns | 10.44 ns | 12.42 ns | 624.5 ns |
| FindSpeciesNearPointNoBranch |              0.2 |            50  | 615.8 ns |  2.12 ns |  1.77 ns | 615.9 ns |
|   FindSpeciesNearPointBranch |              0.2 |           150  | 622.5 ns |  3.96 ns |  3.51 ns | 621.5 ns |
| FindSpeciesNearPointNoBranch |              0.2 |           150  | 618.0 ns |  4.67 ns |  4.37 ns | 616.6 ns |
|   FindSpeciesNearPointBranch |              0.2 |           350  | 620.2 ns |  1.60 ns |  1.33 ns | 620.2 ns |
| FindSpeciesNearPointNoBranch |              0.2 |           350  | 615.3 ns |  2.54 ns |  1.98 ns | 615.4 ns |
|   FindSpeciesNearPointBranch |              0.5 |            50  | 623.7 ns |  4.17 ns |  3.70 ns | 623.1 ns |
| FindSpeciesNearPointNoBranch |              0.5 |            50  | 614.4 ns |  0.80 ns |  0.67 ns | 614.2 ns |
|   FindSpeciesNearPointBranch |              0.5 |           150  | 621.9 ns |  3.68 ns |  3.26 ns | 621.0 ns |
| FindSpeciesNearPointNoBranch |              0.5 |           150  | 616.1 ns |  2.44 ns |  2.28 ns | 615.1 ns |
|   FindSpeciesNearPointBranch |              0.5 |           350  | 623.1 ns |  4.74 ns |  4.21 ns | 621.9 ns |
| FindSpeciesNearPointNoBranch |              0.5 |           350  | 615.7 ns |  1.31 ns |  1.09 ns | 615.1 ns |
```

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeSystem.cs:98]: Thanks for getting the actual numbers. 

I suspect that most of the time the search radius will be quite high, and as there are planned to be roughly 10 species per patch (this is a tweak value in auto-evo), the conclusion is that the branch version is slightly faster usually but sometimes a bit slower. So I suppose this is actually a very slight optimization which means this optimization can be kept.

**Review by hhyyrylainen — APPROVED** [review]
If I read my list of comments right, everything is now in resolved status. So this is ready to merge now. Thanks for sticking with this PR.

Due to my refactoring branch having a bunch of changes I'll likely manually need to merge this there so I'm fine with merging this to master without testing, but later I'd like to ask you to review my branch when that is ready to confirm that I've correctly managed to resolve the merge conflicts.

**Comment by hhyyrylainen** [comment]
Oh right, I just remembered before submitting the review that you'll need to resolve the translation conflicts before I can merge. The usual easy way it to accept the other side changes for everything except en.po and then running the localization update after merging to get everything up to date.

**Pull request merged**
